test/e2e/operator_test.go

```diff
@@ -2723,6 +2723,93 @@ func TestTunableRouterKubeletProbesForCustomIngressController(t *testing.T) {
 	}
 }
 
+// TestIngressControllerServiceNameCollision validates BZ2054200: Don't delete services that are not directly owned by this controller.
+// It creates a service with the same naming convention as the ingress controller creates its own load balancing services.
+// Then it triggers a reconcilation of the ingress operator to see if it will delete our service.
+func TestIngressControllerServiceNameCollision(t *testing.T) {
+	// Create the new private controller that we will later create a service to collide with the naming scheme of this.
+	icName := types.NamespacedName{
+		Namespace: operatorNamespace,
+		Name:      "e2e-name-collision-test",
+	}
+	domain := icName.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newPrivateController(icName, domain)
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller %s: %v", icName, err)
+	}
+
+	// Clean up our new Ingress Controller after we are done.
+	defer assertIngressControllerDeleted(t, kclient, ic)
+
+	// Wait for new ingress controller to come online.
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, icName, availableConditionsForPrivateIngressController...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	// Create services that could collide with the new ingress controller's naming convention for loadbalancer and nodeport.
+	// TRICKY: Our private ingress controller will reconcile (delete) extra loadBalancerServices, even though it isn't one.
+	echoLoadBalancerServiceName := types.NamespacedName{
+		Name:      "router-" + icName.Name,
+		Namespace: "openshift-ingress",
+	}
+	echoLoadBalancerService := buildEchoService(echoLoadBalancerServiceName.Name, echoLoadBalancerServiceName.Namespace, nil)
+	if err := kclient.Create(context.TODO(), echoLoadBalancerService); err != nil {
+		t.Fatalf("failed to create service %s: %v", echoLoadBalancerServiceName, err)
+	}
+
+	echoNodeportServiceName := types.NamespacedName{
+		Name:      "router-nodeport-" + icName.Name,
+		Namespace: "openshift-ingress",
+	}
+	echoNodeportService := buildEchoService(echoNodeportServiceName.Name, echoNodeportServiceName.Namespace, nil)
+	if err := kclient.Create(context.TODO(), echoNodeportService); err != nil {
+		t.Fatalf("failed to create service %s: %v", echoNodeportServiceName, err)
+	}
+
+	// Clean up our new services after we are done.
+	defer func() {
+		if err := kclient.Delete(context.TODO(), echoLoadBalancerService); err != nil {
+			t.Fatalf("failed to delete service %s: %v", echoLoadBalancerServiceName, err)
+		}
```

Comment:
It would be slightly safer to use two separate deferred functions and put the deferred cleanup for the load-balancer service right after the creation of that service (and, more importantly, before the creation of the nodeport service, which could fail and bail out).

Reply:
Agreed. Done.